File: src/main/java/com/indico/query/ModelGroupQuery.java
package com.indico.query;

import com.indico.IndicoClient;
import com.indico.entity.ModelGroup;
import java.util.ArrayList;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ExecutionException;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;

import com.indico.Query;
import com.indico.Async;
import com.indico.ModelGroupGraphQLQuery;
import com.indico.entity.Model;

public class ModelGroupQuery implements Query<ModelGroup> {

    private int id;
    private String name;
    private final IndicoClient indicoClient;

    public ModelGroupQuery(IndicoClient indicoClient) {
        this.indicoClient = indicoClient;
    }

    /**
     * Use to query ModelGroup by id
     *
     * @param id
     * @return ModelGroupQuery
     */
    public ModelGroupQuery id(int id) {
        this.id = id;
        return this;
    }

    /**
     * Use to query ModelGroup by name
     *
     * @param name
     * @return ModelGroupQuery
     */
    public ModelGroupQuery name(String name) {
        this.name = name;
        return this;
    }

    /**
     * Queries the server and returns ModelGroup
     *
     * @return ModelGroup
     */
    @Override
    public ModelGroup query() {
        try{
        ArrayList<Integer> modelGroupIds = new ArrayList<>();
        modelGroupIds.add(this.id);
        ApolloCall<ModelGroupGraphQLQuery.Data> apolloCall = this.indicoClient.apolloClient.query(ModelGroupGraphQLQuery.builder()
                .modelGroupIds(modelGroupIds)
                .build());
        Response<ModelGroupGraphQLQuery.Data> response = Async.executeSync(apolloCall).get();
        ModelGroupGraphQLQuery.ModelGroup mg = response.data().modelGroups().modelGroups().get(0);
        Model model = new Model.Builder()
                .id(mg.selectedModel().id())
                .status(mg.selectedModel().status())
                .build();
        return new ModelGroup.Builder()
                .id(mg.id())
                .name(mg.name())
                .status(mg.status())
                .selectedModel(model)
                .build();
        }catch (CompletionException | ExecutionException | InterruptedException ex){
            throw new RuntimeException("Call to generate the submission result failed", ex);
        }
    }

    /**
     * Refreshes the ModelGroup Object
     *
     * @param obj ModelGroup
     * @return ModelGroup
     */
    @Override
    public ModelGroup refresh(ModelGroup obj) {
        return obj;
    }
}

File: src/main/java/com/indico/query/ModelGroupQuery.kt
package com.indico.query

import com.indico.IndicoClient
import com.indico.entity.Model
import com.indico.entity.ModelGroup
import com.indico.exceptions.IndicoQueryException
import com.indico.graphql.ModelGroupGraphQL
import java.util.ArrayList

class ModelGroupQuery(private val indicoClient: IndicoClient) : Query<ModelGroup?> {
    private var id = 0
    private var name: String? = null

    /**
     * Used to query ModelGroup by id
     *
     * @param id
     * @return ModelGroupQuery
     */
    fun id(id: Int): ModelGroupQuery {
        this.id = id
        return this
    }

    /**
     * Used to query ModelGroup by name
     *
     * @param name
     * @return ModelGroupQuery
     */
    fun name(name: String?): ModelGroupQuery {
        this.name = name
        return this
    }

    /**
     * Queries the server and returns ModelGroup
     *
     * @return ModelGroup
     */
    override fun query(): ModelGroup {
        return try {
            val modelGroupIds = ArrayList<Int>()
            modelGroupIds.add(id)
            val call = ModelGroupGraphQL(ModelGroupGraphQL.Variables(
                modelGroupIds = modelGroupIds
            ))
            val response = indicoClient.execute(call)
            val mg = response.data!!.modelGroups!!.modelGroups!![0]!!
            val model = Model.Builder()
                .id(mg.selectedModel!!.id!!)
                .status(mg.selectedModel.status!!)
                .build()
            ModelGroup.Builder()
                .id(mg.id!!)
                .name(mg.name!!)
                .status(mg.status!!)
                .selectedModel(model)
                .build()
        } catch (ex: RuntimeException) {
            throw IndicoQueryException("Call to generate the submission result failed", ex)
        }
    }

    /**
     * Refreshes the ModelGroup Object
     *
     * @param obj ModelGroup
     * @return ModelGroup
     */
    override fun refresh(obj: ModelGroup?): ModelGroup? {
        return obj
    }

}